File: java/5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 *
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (27.88%)
 * Likes:    4339
 * Dislikes: 396
 * Total Accepted:    655.8K
 * Total Submissions: 2.3M
 * Testcase Example:  '"babad"'
 *
 * Given a string s, find the longest palindromic substring in s. You may
 * assume that the maximum length of s is 1000.
 * 
 * Example 1:
 * 
 * 
 * Input: "babad"
 * Output: "bab"
 * Note: "aba" is also a valid answer.
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "cbbd"
 * Output: "bb"
 * 
 * 
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {

        int len = s.length();
        if (len == 0) {
            return "";
        }
        boolean[][] f = new boolean[len][len];
        int max = 1;
        int start = 0;
        for (int i = 0; i < len; i++) {
            f[i][i] = true;
            if (i + 1 < len && s.charAt(i) == s.charAt(i + 1)) {
                f[i][i + 1] = true;
                start = i;
                max = 2;
            }
        }
        for (int l = 3; l <= len; l++) {
            for (int i = 0; i + l - 1 <= len - 1; i++) {
                int j = i + l - 1;
                if (s.charAt(i) == s.charAt(j) && f[i + 1][j - 1] == true) {
                    f[i][j] = true;
                    start = i;
                    max = l;
                }
            }
        }
        return s.substring(start, start + max);
    }
}
// @lc code=end

File: js/5.longest-palindromic-substring.js
/*
 * @lc app=leetcode id=5 lang=javascript
 *
 * [5] Longest Palindromic Substring
 */
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
	let table = [];
	var start = 0;
	var maxLength = 1;
	// for (let i=0; i<s.length; i++){
	//     let words = [];
	//     for(let j=0; j<s.length;i++){
	//         words.push(false);
	//     }
	//     table.push(words);
	// }
	// };

	for (let i = 0; i < s.length; i++) {
		table[i] = [];
		for (let j = 0; j < s.length; j++) {
			table[i][j] = false;
			// table[i][j] = 0;
		}
	}

	for (let i = 0; i < s.length; i++) {
		// table[i][i] = true;
		table[i][i] = true;
	}

	// for string of 2, compare i, and i+1 th

	for (let i = 0; i < s.length - 1; i++) {
		if (s.charAt(i) === s.charAt(i + 1)) {
			table[i][i + 1] = true;
			start = i;
			maxLength = 2;
		}
	}

	// I don't understand that so much, but just remember

	for (let i = s.length - 1; i >= 0; i--) {
		for (let j = i + 2; j < s.length; j++) {
			table[i][j] = table[i + 1][j - 1] && s.charAt(i) == s.charAt(j);
			console.log('i', i);
			console.log('s.length', s.length);

			if (table[i][j] && j - i + 1 > maxLength) {
				start = i;
				maxLength = j - i + 1;
			}
			console.log('j-i', j - i);
			console.log('\n');
		}
	}

	return s.slice(start, start + maxLength);
};

longestPalindrome('abababbaa');
